src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -692,7 +677,7 @@ private BoundLambda ReallyBind(NamedTypeSymbol delegateType)
             bool reachableEndpoint = ControlFlowPass.Analyze(compilation, lambdaSymbol, block, diagnostics.DiagnosticBag);
             if (reachableEndpoint)
             {
-                if (DelegateNeedsReturn(invokeMethod))
+                if (Binder.MethodOrLambdaRequiresValue(lambdaSymbol, this.Binder.Compilation))
```

Comment:
:memo: Changes to this method are covered by `BuilderFactoryOnMethod_OnLambda_NotTaskLikeTypes`. With this change the test doesn't report `ERR_AnonymousReturnExpected` nor `ERR_CantConvAsyncAnonFuncReturns` below.